### ISSUE TITLE
BKD-2 — `torch.compile` path (`dynamic=False` padded) and `PrecisionConfig` threaded through the blocks

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1566]
**Depends on:** ARCH-4 (#1563), DATA-3 (#1571) · **Blocks:** DEP-2 (#1594)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** Legacy MACE's compilation story is TorchScript: the live blocks carry `torch.jit.is_scripting()` guards (`mace/modules/blocks.py:793`, `mace/modules/utils.py:776`), `@compile_mode("script")` on every block, `@torch.jit.unused` on the hessian helpers (`utils.py:158,189`), and LAMMPS export scripts the model. Its dtype story is hardcoded: `ScaleShiftMACE.forward` computes `total_energy = e0 + inter_e` in the model dtype but routes `node_energy` through `safe_double` (`mace/modules/models.py:585-586`) — a real numeric invariant, but baked into `forward` instead of stated as policy, and *device-conditional*, since `safe_double` is the identity on MPS (`mace/modules/utils.py:22-31`). Static-shape execution exists as three uncoordinated paddings: the ASE calculator's fake-atom padding (`mace/data/padding_tools.py:27-121`, driven from `mace/calculators/mace.py:567-618`), torchsim's fixed pre-filled buffers (`mace/calculators/mace_torchsim.py:308-475`), and LAMMPS ghost slots (`mace/modules/utils.py:786-803`). BKD-1 (#1560) already made every dispatched hot op compile-safe (`custom_op` + `register_fake`, `num_nodes` as SymInt); ARCH-4 (#1563) built the two-phase forward with the derivative seam outside the compiled graph; DATA-3 (#1571) delivers the unified `pad_batch`/`PaddingInfo` contract. This ticket delivers the model-level pieces on top: `torch.compile` as the only compilation path (no TorchScript anywhere on the live path), a `dynamic=False` compiled padded path, a compiled==eager parity smoke, and a declarative `PrecisionConfig` threaded through the block builders that absorbs the `safe_double` invariant.

**Interface & constraints:**

1. **No TorchScript on the live path.** v1 blocks never contain `torch.jit.script`, `@torch.jit.export`, `@torch.jit.unused`, or `torch.jit.is_scripting()` probes, and never trade clarity for scriptability. Deployment export is `torch.export`/AOTInductor and belongs to DEP-2 (#1594); nothing in this ticket may reintroduce a scripting constraint.

2. **Two execution modes over the same model.** Training default is `mode="none"`: dynamic shapes, eager (padding-free). The compiled path is `torch.compile(fullgraph=True, dynamic=False)` over batches produced by DATA-3 (#1571)'s `pad_batch` under a `PaddingPolicy` carrying the legacy budgets (`edge_multiple=64`, `headroom=1.25`, grow-with-one-recompile: exceeding a budget re-pads to a larger budget and triggers exactly one new frame compile). That behaviour is not aspirational — develop already implements it in the ASE calculator (`_auto_estimate_padding`, `mace/calculators/mace.py:567-581`; the bump-and-warn at `:608-618` logs "will trigger one recompile"), and it is pinned by `tests/unit/test_padding.py`. The padding contract is DATA-3 (#1571)'s: *a padded batch is a valid batch* — fake trailing graph, element-0 atoms, self-loop edges at length `2·r_max` where the polynomial cutoff is exactly 0.0 — so the model, the kernels, and the compiled graph are entirely padding-unaware: **no masks in `forward`**, real counts live only in the `PaddingInfo` sidecar, and padding never reorders (`unpad ∘ pad = id` on real slots, the successor of legacy `MACECalculator._slice_real_outputs`, `mace/calculators/mace.py:459`).

3. **Compile hygiene (inherited from BKD-1 (#1560), enforced here at model level).** Each dispatched op is a single opaque node; `num_nodes` is host-side shape metadata (`int` → SymInt), never read from a device tensor or `edge_index.max()`; no data-dependent host reads inside the compiled region (CUDA-graph safety under `reduce-overhead`). The compiled region is the forward per ARCH-4 (#1563)'s design: the strain/derivative seam sits **outside** the compiled graph. No dtype/device/isinstance checks on the hot path — dispatch and precision are resolved at build time, which is what keeps the graph stable. Legacy's own compile path already documents the hazard from the other side: `prepare_graph` skips `requires_grad_` under `torch.compiler.is_compiling()` (`mace/modules/utils.py:805-806`), a branch v1 does not need because the grad leaves are set outside the compiled region.

4. **`PrecisionConfig`** (in `mace_core.config.precision`; torch-free, dtypes are names, never `torch.dtype`):

```python
@dataclass(frozen=True)
class PrecisionPolicy:                  # carried on every op descriptor
    compute_dtype: Precision
    accumulate_dtype: Precision         # a FLOOR, not an exact dtype
    param_dtype: Precision

@dataclass(frozen=True)
class PrecisionConfig:
    interface_dtype: Precision          # dtype of tensors crossing op boundaries
    def policy_for(self, descriptor: OpDescriptor) -> PrecisionPolicy:
        """Authoritative: the policy for THIS dispatched op. For a fused span the
        resolver returns one policy whose accumulate_dtype governs the whole span."""
    @classmethod
    def float64(cls) -> "PrecisionConfig": ...
    @classmethod
    def float32(cls) -> "PrecisionConfig": ...
    @classmethod
    def mixed(cls) -> "PrecisionConfig": ...
        # e.g. linear.compute=bf16/tf32, radial.compute=fp32, reduction.accumulate=fp32
```

   Semantics: precision is defined **per dispatched op or span**, injected through the descriptor and applied at the op boundary — **never a global `torch.autocast`** (fp64-default training makes global autocast wrong, and a fused span has one math dtype, so two `compute_dtype`s inside one kernel are unrepresentable). Inside an op: cast inputs to `policy.compute_dtype`, accumulate reductions in at least `policy.accumulate_dtype`, return in `interface_dtype`. `accumulate_dtype` is a floor: eager torch has no separate accumulator for einsum/matmul-shaped ops, so a backend satisfies the floor either natively (fused kernel with a wider accumulator) or by upcasting compute; on cueq (BKD-3 (#1567)), `accumulate_dtype` wins as the single `math_dtype`. **A floor a device cannot meet degrades, loudly**: fp64 does not exist on MPS, which is precisely why legacy has `safe_double` rather than `.double()`, so the resolver reports the degradation at build time instead of failing at the first cast. Presets seed sensible per-op-kind defaults — linear tolerant of low precision, radial/cutoff high, reductions highest — but the authoritative unit is always the `PrecisionPolicy` on the descriptor.

5. **The reduction invariant, per-quantity.** The legacy `safe_double` at `models.py:586` moves into the reduction policy — `reduction.accumulate_dtype` — and it is **per-quantity**, because P0-6's scale-shift dtype goldens (`tests/unit/test_scale_shift_dtype.py` + references, captured in both fp32 and fp64 for both `MACE` and `ScaleShiftMACE`) pin `total_energy` and `node_energy` as two separate quantities: `total_energy` in the model dtype, `node_energy` in fp64, with the E0-vs-scale-shift application order differing between the two classes — no single dtype policy reproduces both. It is also **per model family**: `MACE` emits both in model dtype (`models.py:399-400`), and so does `MagneticScaleShiftMACE` (`mace/modules/extensions.py:1924-1925`), so the policy is a property of the configured output layer and not a global. Those goldens are the spec this `PrecisionConfig` reduction dtype is written against. Every policy is fixed per dispatch at build time: the hot path has no runtime dtype branching and `register_fake` produces a single-dtype meta.

**Task:**
1. Implement `mace_core.config.precision`: `PrecisionPolicy`, `PrecisionConfig` with `policy_for`, `interface_dtype`, and the `float64`/`float32`/`mixed` presets — no torch/jax imports.
2. Thread it through the v1 model builders: every op descriptor assembled for the ARCH-2 (#1561)/ARCH-3 (#1562)/ARCH-4 (#1563) blocks receives its `PrecisionPolicy` from `PrecisionConfig.policy_for` at build time. No hardcoded `.double()`/`.float()` casts anywhere in a v1 `forward`.
3. Implement the per-quantity reduction policy so `total_energy` (model dtype) and `node_energy` (fp64, degrading on a device without fp64) reproduce the P0-6 goldens for all three output-layer families, at both fp32 and fp64 model dtypes.
4. Wire the compiled padded path: consume DATA-3 (#1571)'s `pad_batch`/`PaddingInfo` under `PaddingPolicy`, compile the forward with `torch.compile(fullgraph=True, dynamic=False)`, keep the ARCH-4 (#1563) derivative seam outside the compiled region, and implement grow-with-one-recompile on budget overflow with the same `edge_multiple`/`headroom` arithmetic the calculator uses today.
5. Add the compiled==eager parity smoke: E/F/stress on real slots, compiled-padded vs eager-unpadded, per-dtype tolerances, across several batch sizes and at least two padding budgets.
6. Add guard tests: no TorchScript symbols on the live path; `mace_core.config.precision` imports neither torch nor jax; no graph break and no silent per-batch recompile.

**Out of scope:** implementing `pad_batch`/`PaddingInfo`/the neighbor-list module (DATA-3 (#1571)); the cueq/oeq backends and their `math_dtype` mapping in practice (BKD-3 (#1567)); LAMMPS/`torch.export` deployment (DEP-2 (#1594)); the op-level conformance suite itself (owned by BKD-1 (#1560) — this ticket adds the model-level smoke, not new suite infrastructure); JAX `jit` scope (which models get jitted is decided elsewhere).

**Acceptance criteria:**
- [ ] `torch.compile(fullgraph=True, dynamic=False)` over the padded path produces no graph break and exactly one frame compile per padding budget; exceeding a budget triggers exactly one recompile at the grown budget.
- [ ] Compiled==eager parity smoke green: E/F/stress on real slots agree within the per-dtype tolerance rows, compiled-padded vs eager-unpadded, for both fp32 and fp64, across ≥3 batch sizes.
- [ ] `PrecisionConfig.float64()` reproduces the P0-6 scale-shift dtype goldens: `total_energy` in the model dtype and `node_energy` in fp64, for the `MACE`, `ScaleShiftMACE` and magnetic output-layer families, in both fp32 and fp64 runs.
- [ ] An `accumulate_dtype` floor the device cannot meet degrades to the model dtype and is reported at build time — matching `safe_double`'s MPS behaviour rather than raising.
- [ ] No `.double()`/`.float()` casts and no `torch.autocast` in any v1 `forward`; no dtype/device branching on the hot path (asserted by test).
- [ ] No `torch.jit.script` / `torch.jit.is_scripting` / `@torch.jit.export` / `@torch.jit.unused` anywhere under `packages/` live-path code (asserted by test).
- [ ] `mace_core.config.precision` imports neither torch nor jax (asserted).
- [ ] A fused span resolves to a single `PrecisionPolicy` whose `accumulate_dtype` governs the whole span (unit-tested via `policy_for`).

**Inventory gaps assigned here:**
- [ ] Calculator padding becomes explicit configuration. It is **not** an uncovered surface any more: `MACE_ASE_PAD_NUM_ATOMS` / `MACE_ASE_PAD_NUM_EDGES` (`mace/calculators/mace.py:423,425`), the fake-graph builder and the real-slot slice are pinned by `tests/unit/test_padding.py` (`TestBuildFakePaddingGraph:27`, `TestSliceRealOutputs:90`). Those cases port onto `PaddingPolicy`/`pad_batch`/`unpad_outputs`; what is still uncovered is the *grow-with-one-recompile* transition, which this ticket adds.

**Verify:**
```bash
python -m pytest packages/mace-core/tests/test_precision_config.py -v                     # policy_for, presets, no torch import
python -m pytest packages/mace-torch/tests/test_compile_padded.py -v                      # fullgraph, dynamic=False, one recompile on grow
python -m pytest packages/mace-torch/tests/test_compiled_eager_parity.py -v               # compiled==eager smoke, real slots
python -m pytest packages/mace-torch/tests/test_scale_shift_dtype_v1.py -v                # vs the P0-6 goldens
```
**Review focus:** the per-quantity reduction dtype against the P0-6 goldens — this is where a silently-wrong energy dtype would hide — that the degradation path matches `safe_double` rather than inventing a new rule, and that the padded compile path introduces no masks, no dtype branches, and no data-dependent host reads into `forward`. CUDA-graph capture (`reduce-overhead`) behaviour is contract-covered, and hardware validation now exists in direction (A100): the pattern captures. What is still unmeasured is the *gain* on a real layer stack — the probe was a four-kernel toy graph.

---
